File: gfg/2024_8_august/python/11_job_sequencing_problem.py
# CODE

class Solution:
    def JobScheduling(self, Jobs, n):
        Jobs.sort(key = lambda x: x.profit, reverse=True)
        
        slots = [False] * (n + 1)
        profit = 0
        cnt = 0
        
        for job in Jobs:
            deadline = job.deadline
            while deadline > 0:
                if not slots[deadline]:
                    slots[deadline] = True
                    cnt += 1
                    profit += job.profit
                    break
                deadline -= 1
                
        return cnt, profit

File: gfg/2024_8_august/c++/11_job_sequencing_problem.cpp
// TIME COMPEXITY
// O(nlogn)

// SPACE COMPEXITY
// O(n)

// CODE

class Solution 
{
    public:
    //Function to find the maximum profit and the number of jobs done.
    static bool cmp(Job a, Job b){
        return a.profit > b.profit;
    }
    vector<int> JobScheduling(Job arr[], int n) 
    { 
        sort(arr, arr+n, cmp);
        int maxDeadline = INT_MIN;
        
        for(int i = 0; i < n; i++){
            maxDeadline = max(maxDeadline,arr[i].dead);
        }
        
        vector<int> deadline(maxDeadline+1,-1);
        int cnt = 0;
        int maxProfit = 0;
        for(int i = 0; i < n; i++){
            int currProfit = arr[i].profit;
            int currId = arr[i].id;
            int currDeadline = arr[i].dead;
            
            for(int j = currDeadline; j > 0; j--){
                if(deadline[j] == -1){
                    cnt++;
                    maxProfit += currProfit;
                    deadline[j] = currId; 
                    break;
                }
            }
        }
        
        vector<int> ans;
        ans.push_back(cnt);
        ans.push_back(maxProfit);
        
        return ans;
        
    } 
};